compiler/rustc_builtin_macros/src/lib.rs

```diff
@@ -131,4 +132,8 @@ pub fn register_builtin_macros(resolver: &mut dyn ResolverExpand) {
 
     let client = proc_macro::bridge::client::Client::expand1(proc_macro::quote);
     register(sym::quote, SyntaxExtensionKind::Bang(Box::new(BangProcMacro { client })));
+    register(
+        sym::contracts_requires,
+        SyntaxExtensionKind::Attr(Box::new(contracts::ExpandRequires)),
+    );
```

Comment:
This macro should be able to use `register_attr!` above.

Reply:
I'm not sure if we actually can, if we want to support arbitrary syntax within the `contracts::require(...)` -- doesn't `register_attr!` mandate that the requires expression conform to `ast::MetaItem`, which imposes restrictions on what the syntax can be, i.e. `x > 0` wouldn't work?

Reply:
I took a closer look at this, and this is very unfortunate. I don't believe the current builtin macro setup allows for "path segments"-like namespacing (like `rustc_contracts::require`). I've tried to change the builtin macro support to allow multiple "segments" via `SmallVec<[Symbol; 2]>`, but as one would expect, that change kept propagating outwards to attributes.

Reply:
sorry for my delay in responding here.

@jieyouxu is exactly right.

specifically, `register_attr!` expands to a `SyntaxExtensionKind::LegacyAttr`, which is where the conformance to `ast::MetaItem` is enforced IIRC. 

The plan here to support code snippets like `x > 0` in a contract form means that we cannot conform to `ast::MetaItem`.

(In theory I could try to extend the `register_attr!` macro to support expansion to non `LegacyAttr`. Is that what you are asking for, @petrochenkov ?)